### DISCUSSION TITLE
how to set rtcm client to connect ngii

### DISCUSSION THREAD

**Discussion by jadebirds** [semuconsulting/PyGPSClient#134] [Q&A]
I'm experiencing an issue with receiving RTCM data and publishing it to window pc. 
Currently, I'm using the pygpsclient and pyrtcm package to receive RTCM data
But there is some problem. 
i can't connect to rtcm server by this code.

Can you give me some solution for this?

<server code>
"""
rtcmsocket.py

A simple example implementation of a GNSS socket reader
using the pyrtcm.RTCMReader iterator functions.

Created on 05 May 2022
@author: semuadmin
"""

import socket
from datetime import datetime
from pyrtcm.rtcmreader import RTCMReader

def read(stream: socket.socket):
    """
    Reads and parses RTCM3 message data from socket stream.
    """

    msgcount = 0
    start = datetime.now()

    rtr = RTCMReader(
        stream,
    )
    try:
        for _, parsed_data in rtr:
            print(parsed_data)
            msgcount += 1
    except KeyboardInterrupt:
        dur = datetime.now() - start
        secs = dur.seconds + dur.microseconds / 1e6
        print("Session terminated by user")
        print(
            f"{msgcount:,d} messages read in {secs:.2f} seconds:",
            f"{msgcount/secs:.2f} msgs per second",
        )

if __name__ == "__main__":
    SERVER = "192.168.0.72"
    PORT = 2101

    print(f"Opening socket {SERVER}:{PORT}...")
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        sock.connect((SERVER, PORT))
        read(sock)

<client code>
"""
rtcm_ntrip_client.py

Illustration of RTCM3 NTRIP Client using GNSSNTRIPClient
class from pygnssutils library. Can be used with any 
NTRIP caster.

NB: requires a valid userid and password. These can be set as
environment variables PYGPSCLIENT_USER and PYGPSCLIENT_PASSWORD,
or passed as keyword arguments user and password.

Usage:

python3 rtcm_ntrip_client.py server="yourcaster" mountpoint="yourmountpoint" user="youruser" password="yourpassword" outfile="rtcmntrip.log"

Run from /examples folder.

Created on 12 Feb 2023

:author: semuadmin
:copyright: SEMU Consulting © 2023
:license: BSD 3-Clause
"""

from os import getenv
from sys import argv
from time import sleep

from pygnssutils import GNSSNTRIPClient

PORT = 2101
HTTPS = 0
if PORT == 443:
    HTTPS = 1

def main(**kwargs):
    """
    Main routine.
    """

    server = kwargs.get("server", "www.gnssdata.or.kr")
    mountpoint = kwargs.get("mountpoint", "GANS-RTCM32")
    user = kwargs.get("user", getenv("PYGPSCLIENT_USER", "mcl20042023@gmail.com"))
    password = kwargs.get("password", getenv("PYGPSCLIENT_PASSWORD", "ngii"))
    outfile = kwargs.get("outfile", "rtcmntrip.log")

    with open(outfile, "wb") as out:
        gnc = GNSSNTRIPClient()

        print(
            f"RTCM NTRIP Client started, writing output to {outfile}... Press CTRL-C to terminate."
        )
        gnc.run(
            server=server,
            port=PORT,
            https=HTTPS,
            mountpoint=mountpoint,
            datatype="RTCM",
            ntripuser=user,
            ntrippassword=password,
            ggainterval=-1,
            output=out,
        )

        try:
            while True:
                sleep(3)
        except KeyboardInterrupt:
            print("RTCM NTRIP Client terminated by User")

if __name__ == "__main__":

    main(**dict(arg.split("=") for arg in argv[1:]))

![스크린샷 2024-05-21 124031](https://github.com/semuconsulting/PyGPSClient/assets/138194761/feb0f2dd-3961-4df2-a7fa-ab0440b46393)

**Comment by semuadmin** [comment]
Hi @jadebirds 

I'm afraid your code snippet above is somewhat garbled so I'm having a hard time working out what you're trying to do here, but an `HTTP/1.1 401 Unauthorised` error simply means that you're using incorrect credentials to access the NTRIP caster. You'll need to verify the correct credentials with your NTRIP service provider.

But `www.gnssdata.or.kr` does not appear to be a valid NTRIP caster? I'm unable to access this URL at all? What is the "192.168.0.72" local address a reference to?
